### PR TITLE
Ability rebuild persistent table free TID list

### PR DESCRIPTION
**Pull request by asimrp**
Each persistent table maintains a list of free tuples.  A tuple is free if and only if `previous_free_tid` is valid (greater than `(0,0)`).  If a tuple is free, the column `persistent_serial_num` indicates its position in the free list.  During normal operation, we maintain the TID and `persistent_serial_num` of the top of the free list in shared memory.  The `previous_free_tid` is used to link free tuples such that it contains the previous free tuple's TID.

If the free list becomes inconsistent (yes, we have seen this happen), this commit introduces one way of reconstructing it.

### CODE REVIEW

**Comment by ivannovick** [comment]
What will be the user's observed behavior of the system if the list becomes inconsistent with this new code?  Thx!

**Comment by asimrp** [comment]
Inconsistent free TID list will no longer cause the recovery to perpetually PANIC.  The free list will be rebuilt automatically if found inconsistent.

**Comment by ivannovick** [comment]
very cool!!!!!!

File: src/backend/cdb/cdbpersistentstore.c [hunk 1]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
+			tbm_add_tuples(tidbm, &entry->key.persistentTid, 1);
+		}
 
 		/*
-		 * Verify the hash table has no free entries left.
+		 * Setup syscache so that PT tuples may be updated using heap
+		 * usual heap access methods.
 		 */
-		while ((entry = hash_seq_search(&iterateStatus)) != NULL)
+		InitCatalogCache();
```
Comment (asimrp) [inline src/backend/cdb/cdbpersistentstore.c]: This is safe because:

1. Catalog cache is backend local.
2. The backend to handle this pass (pass1) of recovery is about to exit.  Rebuilding PT is the last operation performed by this backend.
3. At the beginning of pass 2, we are initializing catalog cache already, see `xlog.c:StartupProcessMain()`.

File: src/backend/cdb/cdbpersistentstore.c [hunk 2]
```diff
@@ -335,69 +348,81 @@ static void PersistentStore_InitScanVerifyFreeEntries(
 		if (!PersistentStore_IsZeroTid(&freeTid))
 		{
 			PersistentStore_DiagnoseDumpTable(storeData, storeSharedData);
-			elog(ERROR, "Expected free TID to be (0,0) when free order number is 0 in '%s'",
```
Comment (jimmyyih) [inline src/backend/cdb/cdbpersistentstore.c]: Should these be ereport instead of elog? :wink:

File: src/backend/cdb/cdbpersistentstore.c [hunk 3]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
-			if (freeOrderNum == 0)
-				break;
-		
 		}
 	}
 
-	if (freeEntryHashTable != NULL)
+	/*
+	 * It is not necessary to validate that sequential search through
+	 * freeEntryHash does not yeild a tuple that is not already seen.
```
Comment (jimmyyih) [inline src/backend/cdb/cdbpersistentstore.c]: s/yeild/yield/

File: src/backend/cdb/cdbpersistentstore.c [hunk 4]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
+				&iterateStatus,
 				freeEntryHashTable);
+		while ((entry = hash_seq_search(&iterateStatus)) != NULL)
+		{
+			tbm_add_tuples(tidbm, &entry->key.persistentTid, 1);
+		}
 
 		/*
-		 * Verify the hash table has no free entries left.
+		 * Setup syscache so that PT tuples may be updated using heap
```
Comment (jimmyyih) [inline src/backend/cdb/cdbpersistentstore.c]: Extra mention of heap here? (unless heap usual heap access methods is a thing)

**Comment by hlinnaka** [comment]
Sounds like a good idea. I'll have to show my ignorance, though, and ask: Why do persistent tables need to maintain a list of free TIDs in the first place?

**Comment by asimrp** [comment]
@hlinnaka, the reason for maintaining a free list of tuples (and not letting vacuum handle freeing) is all inserts/updates to persistent tables are frozen.  Frozen inserts/updates are needed because persistent tables are updated as a post-step of a committed or aborted transaction and the tuples may not have a valid xmin/xmax associated with them.

**Comment by hlinnaka** [comment]
I don't see the issue with freezing. Vacuum runs on the table normally (I presume), even though it won't find any dead tuples in the table, and will make note of free space on pages in the free space map as usual.

If VACUUM doesn't run on this table for some reason, you could still use the usual free space list code to track the free space, by recording the free space there instead of the bespoken free tid list, no?

On 11 May 2016 01:36:24 EEST, Asim RP <notifications@github.com> wrote:
>@hlinnaka, the reason for maintaining a free list of tuples (and not
>letting vacuum handle freeing) is all inserts/updates to persistent
>tables are frozen.  Frozen inserts/updates are needed because
>persistent tables are updated as a post-step of a committed or aborted
>transaction and the tuples may not have a valid xmin/xmax associated
>with them.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/726#issuecomment-218312421

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by asimrp** [comment]
Unless I've got it wrong, it's vacuum that moves used (`LP_USED`) tuples on a page together so that the free space between `pd_lower` and `pd_upper` is increased.  Even if we mark a tuple in persistent tables as dead by setting `LP_DEAD` in the tuple's `ItemIdData`, in order for it to be reused, a vacuum must be run.  This necessity to run vacuum was considered an additional overhead on part of customers and we rejected the option.

**Comment by hlinnaka** [comment]
On 12/05/16 04:14, Asim RP wrote:
> Unless I've got it wrong, it's vacuum that moves used (`LP_USED`)
> tuples on a page together so that the free space between `pd_lower`
> and `pd_upper` is increased.  Even if we mark a tuple in persistent
> tables as dead by setting `LP_DEAD` in the tuple's `ItemIdData`, in
> order for it to be reused, a vacuum must be run.

That's correct.

> This necessity to run vacuum was considered an additional overhead on
> part of customers and we rejected the option.

Running Vacuum itself on a small table that doesn't change that often is 
dirt cheap. And you have to do it for all other catalog tables, anyway. 
I don't see why persistent tables need a different mechanism.

- Heikki

**Comment by ashwinstar** [comment]
Maintaing freelist on top of heap is piece of over-engineering. PT to use native heap free handling and enabling vacuum was one approach considered. The change over to vacuum is involved since currently PT tables are not vacuumed and truncated at all, impact from filerep resync with this change is not known plus has user facing change and action needed to vacuum these tables (are shared tables and not DB specific). free-list seems to brake in some rare scenarios but working otherwise, so rebuilding the same if broken without any user facing change was considered over natively changing the approach.
So, if we have PM buy-in its okay for users to have the change in 4.3_STABLE, would definitely love to apply native heap approach and close the chapter of freeTID once and for all.

**Comment by ivannovick** [comment]
To my understanding of the issue, i think we should first incorporate this change PR to improve current user experience and then also investigate @hlinnaka suggestion to make it even more robust.  I would like to understand better the user facing behavior if we go with the change suggested by @hlinnaka .  In meantime, i think its important to go forward with this current PR that has improvement for existing users soon.

File: src/backend/cdb/cdbpersistentstore.c [hunk 5]
```diff
@@ -335,69 +348,81 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
+					 "Free Entry hashtable corrupted for '%s' (expected free "
+					 "order number " INT64_FORMAT ", entry free order number "
+					 INT64_FORMAT ")",
 				     storeData->tableName,
 				     freeOrderNum,
 				     entry->freeOrderNum);
+				return false;
 			}
-
+			elog(LOG, "%s, " INT64_FORMAT ", %s", ItemPointerToString(&freeTid),
```
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: Is this added for debugging ?

File: src/backend/cdb/cdbpersistentstore.c [hunk 6]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
 			freeOrderNum--;
-
-			if (freeOrderNum == 0)
-				break;
-		
 		}
 	}
 
-	if (freeEntryHashTable != NULL)
+	/*
```
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: This comment is justifying the code removal, but won't be relevant here if read in future. Can it be added at start of Verify function to mention that number of free entries seen to max is checked before calling this function in func ___.

File: src/backend/cdb/cdbpersistentstore.c [hunk 7]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
+	 */
+	Assert(!PersistentStore_IsZeroTid(&storeSharedData->maxTid));
+
+	if (storeSharedData->maxFreeOrderNum < 1)
+	{
+		elog(LOG, "no free tuples in %s, not building any free list",
+			 storeData->tableName);
+		return 0;
+	}
+	elog(LOG, "rebuilding free list in %s with " INT64_FORMAT " tuples",
```
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: s/tuples/free tuples

File: src/backend/cdb/cdbpersistentstore.c (hunk 1 above)
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: Please add this comment to code. Thanks.

File: src/backend/cdb/cdbpersistentstore.c [hunk 8]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
 		 */
-		while ((entry = hash_seq_search(&iterateStatus)) != NULL)
+		InitCatalogCache();
+	}
+	else
+	{
+		/*
+		 * Scan PT to build a tid bitmap of free TIDs.
+		 */
+		elog(LOG, "RebuildFreeList: building freeEntryHashTable first");
```
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: We are not building hashTable, log msg needs correction.

File: src/backend/cdb/cdbpersistentstore.c (hunk 5 above)
Comment (asimrp) [inline src/backend/cdb/cdbpersistentstore.c]: Indeed! I will remove it.

File: src/backend/cdb/cdbpersistentstore.c [hunk 9]
```diff
@@ -415,63 +440,182 @@ static void PersistentStore_InitScanVerifyFreeEntries(
...
+		 storeData->tableName, storeSharedData->maxFreeOrderNum);
+	/*
+	 * Create tid bitmap with enough maxbytes so that it will not be
+	 * lossy.  The bitmap will need as many PageTableEntry objects as
+	 * the number of blocks in PT, which is given by maxTid.  The
+	 * multiplier below to compute the number of bytes needed for as
+	 * many PageTableEntry objects is copied from tbm_create()
+	 * function.
+	 */
+	tidbm = tbm_create(
```
Comment (ashwinstar) [inline src/backend/cdb/cdbpersistentstore.c]: Is it really worth to build tidbm from hashTable and the use it to rebuild. Instead won't creating hashTable for non-recovery case as well better ? Number of freetuples compared to total PT size should be comparatively very small right.

**Comment by jimmyyih** [comment]
Found this warning while compiling this PR's commit on top of our PR:
persistentutil.c:146:1: warning: no previous prototype for function 'gp_persistent_freelist_rebuild' [-Wmissing-prototypes]
gp_persistent_freelist_rebuild(PG_FUNCTION_ARGS)
^
1 warning generated.

**Comment by asimrp** [comment]
Thank you all for the review comments.  Let me note a couple of points upon discussion with Ashwin:

1. It is better to always scan PT for free list and update `previous_free_tid` and `persistent_serial_num` columns in the same scan.  This keeps the code simpler.  And it is more efficient that making two passes, as is currently done by this PR.  I'm going to update the PR with this change soon.

2. Leaving the free space management to vacuum will NOT work.  The only reason being we record `persistentTID` in per-database `gp_relation_node` table.  Every time vacuum moves a tuple to another page, resulting in a different TID, `gp_relation_node` will go out of sync.

**Comment by hlinnaka** [comment]
On 20/05/16 20:56, Asim RP wrote:
> 2. Leaving the free space management to vacuum will NOT work.  The
> only reason being we record `persistentTID` in per-database
> `gp_relation_node` table.  Every time vacuum moves a tuple to another
> page, resulting in a different TID, `gp_relation_node` will go out of
> sync.

We're conflating things. Let's try to get things straight:

1. Lazy VACUUM does not move tuples around. Only VACUUM FULL does.

2. There is currently nothing that prevents you from running "VACUUM 
FULL gp_relation_node". If that breaks things, we should prevent that, 
by adding a check in vacuum.c.

3. Plain database-wide "VACUUM FULL" also vacuums gp_relation_node 
currently. We should prevent that too.

In summary, we should prevent VACUUM FULL from ever running on 
gp_relation_node. Regardless of what we do about free space management. 
But nothing prevents us from doing lazy VACUUM on gp_relation_node, and 
relying on that for free space management, like we do on all other 
catalog tables.

- Heikki

**Comment by asimrp** [comment]
On Fri, May 20, 2016 at 7:05 PM, Heikki Linnakangas <
notifications@github.com> wrote:

>
> In summary, we should prevent VACUUM FULL from ever running on
> gp_relation_node. Regardless of what we do about free space management.
> But nothing prevents us from doing lazy VACUUM on gp_relation_node, and
> relying on that for free space management, like we do on all other
> catalog tables.
>

I am assuming you are referring to shared table
`gp_persistent_relation_node`.  The table `gp_relation_node` is created per
database and refers to in-use tuples from `gp_persistent_relation_node`
using their TIDs.

By preventing VACUUM FULL on `gp_persistent_relation_node`, we are
depriving users from reclaiming disk space.  But that is not a new problem
because today's free list infrastructure too does not allow reclamation of
disk space.  So it makes sense to allow only lazy vacuum on
`gp_persistent_relation_node`.

File: src/backend/access/transam/xlog.c [hunk 10]
```diff
@@ -7020,6 +7020,21 @@ StartupXLOG(void)
...
+		 * heap access methods.  This is safe because:
+		 *
+		 *   1. Catalog cache is backend local.
+		 *
+		 *   2. The backend handling this pass (pass 1) of recovery is
+		 *   about to exit.  Rebuilding PT is the last operation
+		 *   performed by this backend.
+		 *
+		 *   3. At the beginning of pass 2, we are initializing
+		 *   catlaog cache, see StartupProcessMain()
```
Comment (jimmyyih) [inline src/backend/access/transam/xlog.c]: /catlaog/catalog/

**Comment by ashwinstar** [comment]
Looks good 👍

**Comment by hlinnaka** [comment]
On 23/05/16 20:23, Asim RP wrote:
> On Fri, May 20, 2016 at 7:05 PM, Heikki Linnakangas <
> notifications@github.com> wrote:
>
>> In summary, we should prevent VACUUM FULL from ever running on
>> gp_relation_node. Regardless of what we do about free space management.
>> But nothing prevents us from doing lazy VACUUM on gp_relation_node, and
>> relying on that for free space management, like we do on all other
>> catalog tables.
>
> I am assuming you are referring to shared table
> `gp_persistent_relation_node`.  The table `gp_relation_node` is created per
> database and refers to in-use tuples from `gp_persistent_relation_node`
> using their TIDs.

Ah, I see. I had misunderstood how that works.

> By preventing VACUUM FULL on `gp_persistent_relation_node`, we are
> depriving users from reclaiming disk space.  But that is not a new problem
> because today's free list infrastructure too does not allow reclamation of
> disk space.  So it makes sense to allow only lazy vacuum on
> `gp_persistent_relation_node`.

Ok. I'll write a patch for that. Thanks!

- Heikki

**Comment by hlinnaka** [comment]
On 26/05/16 15:47, Heikki Linnakangas wrote:
> On 23/05/16 20:23, Asim RP wrote:
>> On Fri, May 20, 2016 at 7:05 PM, Heikki Linnakangas <
>> notifications@github.com> wrote:
>>
>>> In summary, we should prevent VACUUM FULL from ever running on
>>> gp_relation_node. Regardless of what we do about free space management.
>>> But nothing prevents us from doing lazy VACUUM on gp_relation_node, and
>>> relying on that for free space management, like we do on all other
>>> catalog tables.
>>
>> I am assuming you are referring to shared table
>> `gp_persistent_relation_node`.  The table `gp_relation_node` is created per
>> database and refers to in-use tuples from `gp_persistent_relation_node`
>> using their TIDs.
>
> Ah, I see. I had misunderstood how that works.

To clarify, is it true to say that this query should never return any rows:

select * from gp_relation_node
where persistent_tid not in
   (select ctid from gp_persistent_relation_node);

?

(That query actually fails with "Unexpected internal error" for me at 
the moment, but that seems like an unrelated planner issue. I'll create 
a separate ticket for that..)

- Heikki

**Comment by asimrp** [comment]
@hlinnaka yes, that query should return no tuples.  Sorry for the late response, I somehow missed the notification.

I just rearranged this PR a bit.  The C UDF `gp_persistent_freelist_rebuild()` is now defined as a module under `contrib/gp_internal_tools/`.  Earlier, this was built into postgres binary and loading the binary dynamically failed on RHEL.

File: contrib/gp_internal_tools/Makefile [hunk 11]
```diff
@@ -1,5 +1,6 @@
 MODULE_big = gp_ao_co_diagnostics
 OBJS       = gp_ao_co_diagnostics.o
+MODULES    = gp_persistent_util
```
Comment (asimrp) [inline contrib/gp_internal_tools/Makefile]: I don't understand why but this change is crashing clang's linker.  Travis OSX build failed for this reason.  Here is the relevant snippet from build logs:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C contrib/gp_internal_tools install
gcc -O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-address    -bundle -multiply_defined suppress  gp_ao_co_diagnostics.o -bundle_loader ../../src/backend/postgres  -L../../src/port -L../../src/port -bundle_loader ../../src/backend/postgres  -o gp_ao_co_diagnostics.so
0  0x10d608b51  __assert_rtn + 144
1  0x10d6764cb  ld::tool::OutputFile::compressedOrdinalForAtom(ld::Atom const*) + 281
2  0x10d6771b0  ld::tool::OutputFile::addDyldInfo(ld::Internal&, ld::Internal::FinalSection*, ld::Atom const*, ld::Fixup*, ld::Fixup*, ld::Fixup*, ld::Atom const*, ld::Atom const*, unsigned long long, unsigned long long) + 2364
3  0x10d66cfdb  ld::tool::OutputFile::generateLinkEditInfo(ld::Internal&) + 1281
4  0x10d668530  ld::tool::OutputFile::write(ld::Internal&) + 116
5  0x10d60981a  main + 1028
A linker snapshot was created at:
	/tmp/gp_ao_co_diagnostics.so-2016-05-02-223301.ld-snapshot
ld: Assertion failed: (0 && "dylib not assigned ordinal"), function compressedOrdinalForAtom, file /SourceCache/ld64/ld64-241.9/src/ld/OutputFile.cpp, line 3495.
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [gp_ao_co_diagnostics.so] Error 1
make: *** [install] Error 2
```

Note that GCC compiles this fine.  Will hold this PR until I figure out what's going on.  I think moving `gp_persistent_util.c` to a separate directory under `contrib/` may fix the issue.

**Comment by asimrp** [comment]
Pushed to master, thanks for the review.
